File: packages/less/src/less/environment/environment.test.js
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import Environment from './environment.js';
import logger from '../logger.js';

// Mock logger
vi.mock('../logger.js', () => ({
    default: {
        warn: vi.fn()
    }
}));

describe('Environment', () => {
    let mockFileManager;
    let mockFileManager2;
    let mockPluginManager;

    beforeEach(() => {
        vi.clearAllMocks();
        
        mockFileManager = {
            supports: vi.fn(),
            supportsSync: vi.fn()
        };
        
        mockFileManager2 = {
            supports: vi.fn(),
            supportsSync: vi.fn()
        };

        mockPluginManager = {
            getFileManagers: vi.fn(() => [mockFileManager2])
        };
    });

    afterEach(() => {
        vi.resetAllMocks();
    });

    describe('constructor', () => {
        it('should create an Environment with empty fileManagers when none provided', () => {
            const env = new Environment();
            expect(env.fileManagers).toEqual([]);
        });

        it('should create an Environment with provided fileManagers', () => {
            const fileManagers = [mockFileManager];
            const env = new Environment(null, fileManagers);
            expect(env.fileManagers).toBe(fileManagers);
        });

        it('should handle null externalEnvironment', () => {
            const env = new Environment(null);
            expect(env.fileManagers).toEqual([]);
        });

        it('should handle undefined externalEnvironment', () => {
            const env = new Environment(undefined);
            expect(env.fileManagers).toEqual([]);
        });

        it('should bind optional functions from externalEnvironment', () => {
            const mockEncodeBase64 = vi.fn();
            const mockMimeLookup = vi.fn();
            const mockCharsetLookup = vi.fn();
            const mockGetSourceMapGenerator = vi.fn();
            
            const externalEnv = {
                encodeBase64: mockEncodeBase64,
                mimeLookup: mockMimeLookup,
                charsetLookup: mockCharsetLookup,
                getSourceMapGenerator: mockGetSourceMapGenerator,
                someOtherProp: 'ignored'
            };

            const env = new Environment(externalEnv);

            expect(env.encodeBase64).toBeDefined();
            expect(env.mimeLookup).toBeDefined();
            expect(env.charsetLookup).toBeDefined();
            expect(env.getSourceMapGenerator).toBeDefined();
            expect(env.someOtherProp).toBeUndefined();
        });

        it('should bind functions with correct context', () => {
            const externalEnv = {
                testValue: 'test',
                encodeBase64: function() {
                    return this.testValue;
                }
            };

            const env = new Environment(externalEnv);
            expect(env.encodeBase64()).toBe('test');
        });

        it('should handle missing optional functions gracefully', () => {
            const externalEnv = {
                encodeBase64: vi.fn()
                // missing mimeLookup, charsetLookup, getSourceMapGenerator
            };

            const env = new Environment(externalEnv);

            expect(env.encodeBase64).toBeDefined();
            expect(env.mimeLookup).toBeUndefined();
            expect(env.charsetLookup).toBeUndefined();
            expect(env.getSourceMapGenerator).toBeUndefined();
        });

        it('should not warn for missing optional functions', () => {
            new Environment({});
            expect(logger.warn).not.toHaveBeenCalled();
        });

        it('should handle empty externalEnvironment object', () => {
            const env = new Environment({});
            expect(env.encodeBase64).toBeUndefined();
            expect(env.mimeLookup).toBeUndefined();
            expect(env.charsetLookup).toBeUndefined();
            expect(env.getSourceMapGenerator).toBeUndefined();
        });
    });

    describe('getFileManager', () => {
        let env;

        beforeEach(() => {
            env = new Environment(null, [mockFileManager]);
        });

        it('should warn when filename is not provided', () => {
            env.getFileManager(undefined, '/path', {});
            expect(logger.warn).toHaveBeenCalledWith('getFileManager called with no filename.. Please report this issue. continuing.');
        });

        it('should warn when filename is empty string', () => {
            env.getFileManager('', '/path', {});
            expect(logger.warn).toHaveBeenCalledWith('getFileManager called with no filename.. Please report this issue. continuing.');
        });

        it('should warn when currentDirectory is undefined', () => {
            env.getFileManager('test.less', undefined, {});
            expect(logger.warn).toHaveBeenCalledWith('getFileManager called with null directory.. Please report this issue. continuing.');
        });

        it('should not warn when currentDirectory is null', () => {
            mockFileManager.supports.mockReturnValue(true);
            env.getFileManager('test.less', null, {});
            expect(logger.warn).toHaveBeenCalledTimes(0);
        });

        it('should not warn when currentDirectory is empty string', () => {
            mockFileManager.supports.mockReturnValue(true);
            env.getFileManager('test.less', '', {});
            expect(logger.warn).toHaveBeenCalledTimes(0);
        });

        it('should return file manager that supports the file (async)', () => {
            mockFileManager.supports.mockReturnValue(true);
            
            const result = env.getFileManager('test.less', '/path', {}, {}, false);
            
            expect(mockFileManager.supports).toHaveBeenCalledWith('test.less', '/path', {}, {});
            expect(result).toBe(mockFileManager);
        });

        it('should return file manager that supports the file (sync)', () => {
            mockFileManager.supportsSync.mockReturnValue(true);
            
            const result = env.getFileManager('test.less', '/path', {}, {}, true);
            
            expect(mockFileManager.supportsSync).toHaveBeenCalledWith('test.less', '/path', {}, {});
            expect(result).toBe(mockFileManager);
        });

        it('should return null when no file manager supports the file', () => {
            mockFileManager.supports.mockReturnValue(false);
            
            const result = env.getFileManager('test.less', '/path', {}, {}, false);
            
            expect(result).toBe(null);
        });

        it('should check file managers in reverse order', () => {
            const fileManager1 = { supports: vi.fn(() => false) };
            const fileManager2 = { supports: vi.fn(() => true) };
            const fileManager3 = { supports: vi.fn(() => true) };
            
            env.fileManagers = [fileManager1, fileManager2, fileManager3];
            
            const result = env.getFileManager('test.less', '/path', {}, {}, false);
            
            // Should check in reverse order and return the last one that supports
            expect(fileManager3.supports).toHaveBeenCalled();
            expect(fileManager2.supports).not.toHaveBeenCalled();
            expect(fileManager1.supports).not.toHaveBeenCalled();
            expect(result).toBe(fileManager3);
        });

        it('should include plugin manager file managers when pluginManager is provided', () => {
            const options = { pluginManager: mockPluginManager };
            mockFileManager2.supports.mockReturnValue(true);
            
            const result = env.getFileManager('test.less', '/path', options, {}, false);
            
            expect(mockPluginManager.getFileManagers).toHaveBeenCalled();
            expect(mockFileManager2.supports).toHaveBeenCalledWith('test.less', '/path', options, {});
            expect(result).toBe(mockFileManager2);
        });

        it('should prioritize plugin manager file managers over environment file managers', () => {
            const options = { pluginManager: mockPluginManager };
            mockFileManager.supports.mockReturnValue(true);
            mockFileManager2.supports.mockReturnValue(true);
            
            const result = env.getFileManager('test.less', '/path', options, {}, false);
            
            // Plugin manager file managers come after environment file managers,
            // so they're checked first (reverse order)
            expect(mockFileManager2.supports).toHaveBeenCalled();
            expect(mockFileManager.supports).not.toHaveBeenCalled();
            expect(result).toBe(mockFileManager2);
        });

        it('should handle empty plugin manager file managers', () => {
            const emptyPluginManager = { getFileManagers: vi.fn(() => []) };
            const options = { pluginManager: emptyPluginManager };
            mockFileManager.supports.mockReturnValue(true);
            
            const result = env.getFileManager('test.less', '/path', options, {}, false);
            
            expect(result).toBe(mockFileManager);
        });

        it('should handle options without pluginManager', () => {
            const options = {};
            mockFileManager.supports.mockReturnValue(true);
            
            const result = env.getFileManager('test.less', '/path', options, {}, false);
            
            expect(result).toBe(mockFileManager);
        });

        it('should handle null options', () => {
            mockFileManager.supports.mockReturnValue(true);
            
            // The original code has a bug - it doesn't handle null options
            // This test documents the current behavior
            expect(() => {
                env.getFileManager('test.less', '/path', null, {}, false);
            }).toThrow();
        });

        it('should handle undefined options', () => {
            mockFileManager.supports.mockReturnValue(true);
            
            // The original code has a bug - it doesn't handle undefined options
            // This test documents the current behavior
            expect(() => {
                env.getFileManager('test.less', '/path', undefined, {}, false);
            }).toThrow();
        });

        it('should pass all parameters to supports method', () => {
            const filename = 'test.less';
            const currentDirectory = '/some/path';
            const options = { compress: true };
            const environment = { debug: true };
            
            mockFileManager.supports.mockReturnValue(true);
            
            env.getFileManager(filename, currentDirectory, options, environment, false);
            
            expect(mockFileManager.supports).toHaveBeenCalledWith(filename, currentDirectory, options, environment);
        });

        it('should pass all parameters to supportsSync method', () => {
            const filename = 'test.less';
            const currentDirectory = '/some/path';
            const options = { compress: true };
            const environment = { debug: true };
            
            mockFileManager.supportsSync.mockReturnValue(true);
            
            env.getFileManager(filename, currentDirectory, options, environment, true);
            
            expect(mockFileManager.supportsSync).toHaveBeenCalledWith(filename, currentDirectory, options, environment);
        });
    });

    describe('addFileManager', () => {
        it('should add file manager to the list', () => {
            const env = new Environment();
            expect(env.fileManagers).toHaveLength(0);
            
            env.addFileManager(mockFileManager);
            
            expect(env.fileManagers).toHaveLength(1);
            expect(env.fileManagers[0]).toBe(mockFileManager);
        });

        it('should add multiple file managers', () => {
            const env = new Environment();
            
            env.addFileManager(mockFileManager);
            env.addFileManager(mockFileManager2);
            
            expect(env.fileManagers).toHaveLength(2);
            expect(env.fileManagers[0]).toBe(mockFileManager);
            expect(env.fileManagers[1]).toBe(mockFileManager2);
        });

        it('should add file manager to existing list', () => {
            const env = new Environment(null, [mockFileManager]);
            expect(env.fileManagers).toHaveLength(1);
            
            env.addFileManager(mockFileManager2);
            
            expect(env.fileManagers).toHaveLength(2);
            expect(env.fileManagers[0]).toBe(mockFileManager);
            expect(env.fileManagers[1]).toBe(mockFileManager2);
        });

        it('should handle null file manager', () => {
            const env = new Environment();
            
            env.addFileManager(null);
            
            expect(env.fileManagers).toHaveLength(1);
            expect(env.fileManagers[0]).toBe(null);
        });

        it('should handle undefined file manager', () => {
            const env = new Environment();
            
            env.addFileManager(undefined);
            
            expect(env.fileManagers).toHaveLength(1);
            expect(env.fileManagers[0]).toBe(undefined);
        });
    });

    describe('clearFileManagers', () => {
        it('should clear empty file managers list', () => {
            const env = new Environment();
            
            env.clearFileManagers();
            
            expect(env.fileManagers).toEqual([]);
        });

        it('should clear file managers list with items', () => {
            const env = new Environment(null, [mockFileManager, mockFileManager2]);
            expect(env.fileManagers).toHaveLength(2);
            
            env.clearFileManagers();
            
            expect(env.fileManagers).toEqual([]);
        });

        it('should allow adding file managers after clearing', () => {
            const env = new Environment(null, [mockFileManager]);
            
            env.clearFileManagers();
            env.addFileManager(mockFileManager2);
            
            expect(env.fileManagers).toHaveLength(1);
            expect(env.fileManagers[0]).toBe(mockFileManager2);
        });
    });

    describe('integration tests', () => {
        it('should work with complete workflow', () => {
            const externalEnv = {
                encodeBase64: vi.fn(() => 'encoded'),
                mimeLookup: vi.fn(() => 'text/css')
            };
            
            const env = new Environment(externalEnv, [mockFileManager]);
            
            // Test external environment functions
            expect(env.encodeBase64()).toBe('encoded');
            expect(env.mimeLookup()).toBe('text/css');
            
            // Test file manager operations
            env.addFileManager(mockFileManager2);
            expect(env.fileManagers).toHaveLength(2);
            
            // Test getFileManager
            mockFileManager2.supports.mockReturnValue(true);
            const result = env.getFileManager('test.less', '/path', {}, {}, false);
            expect(result).toBe(mockFileManager2);
            
            // Test clear
            env.clearFileManagers();
            expect(env.fileManagers).toHaveLength(0);
        });

        it('should handle complex plugin manager scenario', () => {
            const pluginFileManager1 = { supports: vi.fn(() => false) };
            const pluginFileManager2 = { supports: vi.fn(() => true) };
            const complexPluginManager = {
                getFileManagers: vi.fn(() => [pluginFileManager1, pluginFileManager2])
            };
            
            const env = new Environment(null, [mockFileManager]);
            mockFileManager.supports.mockReturnValue(true);
            
            const options = { pluginManager: complexPluginManager };
            const result = env.getFileManager('test.less', '/path', options, {}, false);
            
            // Should return the plugin file manager that supports (checked first due to reverse order)
            expect(result).toBe(pluginFileManager2);
            expect(pluginFileManager2.supports).toHaveBeenCalled();
            expect(pluginFileManager1.supports).not.toHaveBeenCalled();
            expect(mockFileManager.supports).not.toHaveBeenCalled();
        });
    });
});

File: packages/less/src/less/environment/environment_test.go
package environment

import (
	"testing"

	"github.com/toakleaf/less.go/packages/less/src/less"
)

// MockFileManager implements FileManager for testing
type MockFileManager struct {
	SupportsFunc     func(string, string, map[string]any, map[string]any) bool
	SupportsSyncFunc func(string, string, map[string]any, map[string]any) bool
}

func (m *MockFileManager) Supports(filename, currentDirectory string, options map[string]any, environment map[string]any) bool {
	if m.SupportsFunc != nil {
		return m.SupportsFunc(filename, currentDirectory, options, environment)
	}
	return false
}

func (m *MockFileManager) SupportsSync(filename, currentDirectory string, options map[string]any, environment map[string]any) bool {
	if m.SupportsSyncFunc != nil {
		return m.SupportsSyncFunc(filename, currentDirectory, options, environment)
	}
	return false
}

// MockPluginManager implements PluginManager for testing
type MockPluginManager struct {
	GetFileManagersFunc func() []FileManager
}

func (m *MockPluginManager) GetFileManagers() []FileManager {
	if m.GetFileManagersFunc != nil {
		return m.GetFileManagersFunc()
	}
	return []FileManager{}
}

// MockLogListener implements LogListener for testing warnings
type MockLogListener struct {
	WarnMessages  []any
	ErrorMessages []any
	InfoMessages  []any
	DebugMessages []any
}

func (m *MockLogListener) Error(msg any) {
	m.ErrorMessages = append(m.ErrorMessages, msg)
}

func (m *MockLogListener) Warn(msg any) {
	m.WarnMessages = append(m.WarnMessages, msg)
}

func (m *MockLogListener) Info(msg any) {
	m.InfoMessages = append(m.InfoMessages, msg)
}

func (m *MockLogListener) Debug(msg any) {
	m.DebugMessages = append(m.DebugMessages, msg)
}

func TestEnvironmentConstructor(t *testing.T) {
	t.Run("should create an Environment with empty fileManagers when none provided", func(t *testing.T) {
		env := NewEnvironment(nil, nil)
		if len(env.FileManagers) != 0 {
			t.Errorf("Expected empty FileManagers, got %d", len(env.FileManagers))
		}
	})

	t.Run("should create an Environment with provided fileManagers", func(t *testing.T) {
		mockFileManager := &MockFileManager{}
		fileManagers := []FileManager{mockFileManager}
		env := NewEnvironment(nil, fileManagers)
		if len(env.FileManagers) != 1 {
			t.Errorf("Expected 1 FileManager, got %d", len(env.FileManagers))
		}
		if env.FileManagers[0] != mockFileManager {
			t.Error("FileManager was not set correctly")
		}
	})

	t.Run("should handle nil externalEnvironment", func(t *testing.T) {
		env := NewEnvironment(nil, nil)
		if len(env.FileManagers) != 0 {
			t.Errorf("Expected empty FileManagers, got %d", len(env.FileManagers))
		}
	})

	t.Run("should bind optional functions from externalEnvironment", func(t *testing.T) {
		externalEnv := map[string]any{
			"encodeBase64":          func() string { return "encoded" },
			"mimeLookup":            func() string { return "text/css" },
			"charsetLookup":         func() string { return "utf-8" },
			"getSourceMapGenerator": func() any { return "sourcemap" },
			"someOtherProp":         "ignored",
		}

		env := NewEnvironment(externalEnv, nil)

		if env.EncodeBase64 == nil {
			t.Error("EncodeBase64 should be defined")
		}
		if env.MimeLookup == nil {
			t.Error("MimeLookup should be defined")
		}
		if env.CharsetLookup == nil {
			t.Error("CharsetLookup should be defined")
		}
		if env.GetSourceMapGenerator == nil {
			t.Error("GetSourceMapGenerator should be defined")
		}
	})

	t.Run("should bind functions with correct context", func(t *testing.T) {
		externalEnv := map[string]any{
			"encodeBase64": func() string { return "test" },
		}

		env := NewEnvironment(externalEnv, nil)
		if env.EncodeBase64() != "test" {
			t.Errorf("Expected 'test', got '%s'", env.EncodeBase64())
		}
	})

	t.Run("should handle missing optional functions gracefully", func(t *testing.T) {
		externalEnv := map[string]any{
			"encodeBase64": func() string { return "encoded" },
		}

		env := NewEnvironment(externalEnv, nil)

		if env.EncodeBase64 == nil {
			t.Error("EncodeBase64 should be defined")
		}
		if env.MimeLookup != nil {
			t.Error("MimeLookup should be nil")
		}
		if env.CharsetLookup != nil {
			t.Error("CharsetLookup should be nil")
		}
		if env.GetSourceMapGenerator != nil {
			t.Error("GetSourceMapGenerator should be nil")
		}
	})

	t.Run("should not warn for missing optional functions", func(t *testing.T) {
		mockListener := &MockLogListener{}
		less.AddListener(mockListener)
		defer less.RemoveListener(mockListener)

		NewEnvironment(map[string]any{}, nil)

		if len(mockListener.WarnMessages) > 0 {
			t.Error("Should not warn for missing optional functions")
		}
	})

	t.Run("should handle empty externalEnvironment object", func(t *testing.T) {
		env := NewEnvironment(map[string]any{}, nil)
		if env.EncodeBase64 != nil {
			t.Error("EncodeBase64 should be nil")
		}
		if env.MimeLookup != nil {
			t.Error("MimeLookup should be nil")
		}
		if env.CharsetLookup != nil {
			t.Error("CharsetLookup should be nil")
		}
		if env.GetSourceMapGenerator != nil {
			t.Error("GetSourceMapGenerator should be nil")
		}
	})
}

func TestGetFileManager(t *testing.T) {
	var mockFileManager *MockFileManager
	var env *Environment

	setUp := func() {
		mockFileManager = &MockFileManager{}
		env = NewEnvironment(nil, []FileManager{mockFileManager})
	}

	t.Run("should warn when filename is not provided", func(t *testing.T) {
		setUp()
		mockListener := &MockLogListener{}
		less.AddListener(mockListener)
		defer less.RemoveListener(mockListener)

		env.GetFileManager("", "/path", map[string]any{}, map[string]any{}, false)

		if len(mockListener.WarnMessages) == 0 {
			t.Error("Should warn when filename is empty")
		}
		expectedMsg := "getFileManager called with no filename.. Please report this issue. continuing."
		if mockListener.WarnMessages[0] != expectedMsg {
			t.Errorf("Expected warning message '%s', got '%v'", expectedMsg, mockListener.WarnMessages[0])
		}
	})

	t.Run("should warn when currentDirectory is nil", func(t *testing.T) {
		setUp()
		mockListener := &MockLogListener{}
		less.AddListener(mockListener)
		defer less.RemoveListener(mockListener)

		env.GetFileManager("test.less", "", map[string]any{}, map[string]any{}, false)

		if len(mockListener.WarnMessages) == 0 {
			t.Error("Should warn when currentDirectory is nil")
		}
		expectedMsg := "getFileManager called with null directory.. Please report this issue. continuing."
		if mockListener.WarnMessages[0] != expectedMsg {
			t.Errorf("Expected warning message '%s', got '%v'", expectedMsg, mockListener.WarnMessages[0])
		}
	})

	t.Run("should not warn when currentDirectory is empty string", func(t *testing.T) {
		setUp()
		mockListener := &MockLogListener{}
		less.AddListener(mockListener)
		defer less.RemoveListener(mockListener)

		mockFileManager.SupportsFunc = func(string, string, map[string]any, map[string]any) bool { return true }
		env.GetFileManager("test.less", "null", map[string]any{}, map[string]any{}, false)

		if len(mockListener.WarnMessages) > 0 {
			t.Error("Should not warn when currentDirectory is empty string")
		}
	})

	t.Run("should return file manager that supports the file (async)", func(t *testing.T) {
		setUp()
		mockFileManager.SupportsFunc = func(filename, currentDirectory string, options, environment map[string]any) bool {
			return filename == "test.less" && currentDirectory == "/path"
		}

		result := env.GetFileManager("test.less", "/path", map[string]any{}, map[string]any{}, false)

		if result != mockFileManager {
			t.Error("Should return the mockFileManager")
		}
	})

	t.Run("should return file manager that supports the file (sync)", func(t *testing.T) {
		setUp()
		mockFileManager.SupportsSyncFunc = func(filename, currentDirectory string, options, environment map[string]any) bool {
			return filename == "test.less" && currentDirectory == "/path"
		}

		result := env.GetFileManager("test.less", "/path", map[string]any{}, map[string]any{}, true)

		if result != mockFileManager {
			t.Error("Should return the mockFileManager")
		}
	})

	t.Run("should return nil when no file manager supports the file", func(t *testing.T) {
		setUp()
		mockFileManager.SupportsFunc = func(string, string, map[string]any, map[string]any) bool { return false }

		result := env.GetFileManager("test.less", "/path", map[string]any{}, map[string]any{}, false)

		if result != nil {
			t.Error("Should return nil when no file manager supports the file")
		}
	})

	t.Run("should check file managers in reverse order", func(t *testing.T) {
		setUp()
		fileManager1 := &MockFileManager{
			SupportsFunc: func(string, string, map[string]any, map[string]any) bool { return false },
		}
		fileManager2 := &MockFileManager{
			SupportsFunc: func(string, string, map[string]any, map[string]any) bool { return true },
		}
		fileManager3 := &MockFileManager{
			SupportsFunc: func(string, string, map[string]any, map[string]any) bool { return true },
		}

		env.FileManagers = []FileManager{fileManager1, fileManager2, fileManager3}

		result := env.GetFileManager("test.less", "/path", map[string]any{}, map[string]any{}, false)

		// Should return the last one that supports (checked first due to reverse order)
		if result != fileManager3 {
			t.Error("Should return fileManager3 (last one checked)")
		}
	})

	t.Run("should include plugin manager file managers when pluginManager is provided", func(t *testing.T) {
		setUp()
		mockFileManager2 := &MockFileManager{
			SupportsFunc: func(string, string, map[string]any, map[string]any) bool { return true },
		}
		mockPluginManager := &MockPluginManager{
			GetFileManagersFunc: func() []FileManager { return []FileManager{mockFileManager2} },
		}

		options := map[string]any{"pluginManager": mockPluginManager}
		result := env.GetFileManager("test.less", "/path", options, map[string]any{}, false)

		if result != mockFileManager2 {
			t.Error("Should return mockFileManager2 from plugin manager")
		}
	})

	t.Run("should prioritize plugin manager file managers over environment file managers", func(t *testing.T) {
		setUp()
		mockFileManager.SupportsFunc = func(string, string, map[string]any, map[string]any) bool { return true }
		
		mockFileManager2 := &MockFileManager{
			SupportsFunc: func(string, string, map[string]any, map[string]any) bool { return true },
		}
		mockPluginManager := &MockPluginManager{
			GetFileManagersFunc: func() []FileManager { return []FileManager{mockFileManager2} },
		}

		options := map[string]any{"pluginManager": mockPluginManager}
		result := env.GetFileManager("test.less", "/path", options, map[string]any{}, false)

		// Plugin manager file managers come after environment file managers,
		// so they're checked first (reverse order)
		if result != mockFileManager2 {
			t.Error("Should return mockFileManager2 from plugin manager (higher priority)")
		}
	})

	t.Run("should handle empty plugin manager file managers", func(t *testing.T) {
		setUp()
		mockFileManager.SupportsFunc = func(string, string, map[string]any, map[string]any) bool { return true }
		
		emptyPluginManager := &MockPluginManager{
			GetFileManagersFunc: func() []FileManager { return []FileManager{} },
		}

		options := map[string]any{"pluginManager": emptyPluginManager}
		result := env.GetFileManager("test.less", "/path", options, map[string]any{}, false)

		if result != mockFileManager {
			t.Error("Should return environment file manager when plugin manager is empty")
		}
	})

	t.Run("should handle options without pluginManager", func(t *testing.T) {
		setUp()
		mockFileManager.SupportsFunc = func(string, string, map[string]any, map[string]any) bool { return true }

		options := map[string]any{}
		result := env.GetFileManager("test.less", "/path", options, map[string]any{}, false)

		if result != mockFileManager {
			t.Error("Should return environment file manager")
		}
	})

	t.Run("should handle nil options", func(t *testing.T) {
		setUp()
		mockFileManager.SupportsFunc = func(string, string, map[string]any, map[string]any) bool { return true }

		// The original JavaScript code has a bug - it doesn't handle null options
		// This test documents the current behavior - it should panic
		defer func() {
			if r := recover(); r == nil {
				t.Error("Expected panic when options is nil, but no panic occurred")
			}
		}()

		env.GetFileManager("test.less", "/path", nil, map[string]any{}, false)
	})

	t.Run("should handle undefined options", func(t *testing.T) {
		setUp()
		mockFileManager.SupportsFunc = func(string, string, map[string]any, map[string]any) bool { return true }

		// The original JavaScript code has a bug - it doesn't handle undefined options
		// This test documents the current behavior - it should panic
		defer func() {
			if r := recover(); r == nil {
				t.Error("Expected panic when options is undefined, but no panic occurred")
			}
		}()

		env.GetFileManager("test.less", "/path", nil, map[string]any{}, false)
	})

	t.Run("should pass all parameters to supports method", func(t *testing.T) {
		setUp()
		var capturedFilename, capturedCurrentDirectory string
		var capturedOptions, capturedEnvironment map[string]any

		mockFileManager.SupportsFunc = func(filename, currentDirectory string, options, environment map[string]any) bool {
			capturedFilename = filename
			capturedCurrentDirectory = currentDirectory
			capturedOptions = options
			capturedEnvironment = environment
			return true
		}

		filename := "test.less"
		currentDirectoryStr := "/some/path"
		options := map[string]any{"compress": true}
		environment := map[string]any{"debug": true}

		env.GetFileManager(filename, currentDirectoryStr, options, environment, false)

		if capturedFilename != filename {
			t.Errorf("Expected filename '%s', got '%s'", filename, capturedFilename)
		}
		if capturedCurrentDirectory != currentDirectoryStr {
			t.Errorf("Expected currentDirectory '%s', got '%s'", currentDirectoryStr, capturedCurrentDirectory)
		}
		if capturedOptions["compress"] != true {
			t.Error("Options not passed correctly")
		}
		if capturedEnvironment["debug"] != true {
			t.Error("Environment not passed correctly")
		}
	})

	t.Run("should pass all parameters to supportsSync method", func(t *testing.T) {
		setUp()
		var capturedFilename, capturedCurrentDirectory string
		var capturedOptions, capturedEnvironment map[string]any

		mockFileManager.SupportsSyncFunc = func(filename, currentDirectory string, options, environment map[string]any) bool {
			capturedFilename = filename
			capturedCurrentDirectory = currentDirectory
			capturedOptions = options
			capturedEnvironment = environment
			return true
		}

		filename := "test.less"
		currentDirectoryStr := "/some/path"
		options := map[string]any{"compress": true}
		environment := map[string]any{"debug": true}

		env.GetFileManager(filename, currentDirectoryStr, options, environment, true)

		if capturedFilename != filename {
			t.Errorf("Expected filename '%s', got '%s'", filename, capturedFilename)
		}
		if capturedCurrentDirectory != currentDirectoryStr {
			t.Errorf("Expected currentDirectory '%s', got '%s'", currentDirectoryStr, capturedCurrentDirectory)
		}
		if capturedOptions["compress"] != true {
			t.Error("Options not passed correctly")
		}
		if capturedEnvironment["debug"] != true {
			t.Error("Environment not passed correctly")
		}
	})
}

func TestAddFileManager(t *testing.T) {
	t.Run("should add file manager to the list", func(t *testing.T) {
		env := NewEnvironment(nil, nil)
		if len(env.FileManagers) != 0 {
			t.Error("Expected empty FileManagers initially")
		}

		mockFileManager := &MockFileManager{}
		env.AddFileManager(mockFileManager)

		if len(env.FileManagers) != 1 {
			t.Errorf("Expected 1 FileManager, got %d", len(env.FileManagers))
		}
		if env.FileManagers[0] != mockFileManager {
			t.Error("FileManager not added correctly")
		}
	})

	t.Run("should add multiple file managers", func(t *testing.T) {
		env := NewEnvironment(nil, nil)
		mockFileManager := &MockFileManager{}
		mockFileManager2 := &MockFileManager{}

		env.AddFileManager(mockFileManager)
		env.AddFileManager(mockFileManager2)

		if len(env.FileManagers) != 2 {
			t.Errorf("Expected 2 FileManagers, got %d", len(env.FileManagers))
		}
		if env.FileManagers[0] != mockFileManager {
			t.Error("First FileManager not correct")
		}
		if env.FileManagers[1] != mockFileManager2 {
			t.Error("Second FileManager not correct")
		}
	})

	t.Run("should add file manager to existing list", func(t *testing.T) {
		mockFileManager := &MockFileManager{}
		env := NewEnvironment(nil, []FileManager{mockFileManager})
		if len(env.FileManagers) != 1 {
			t.Error("Expected 1 FileManager initially")
		}

		mockFileManager2 := &MockFileManager{}
		env.AddFileManager(mockFileManager2)

		if len(env.FileManagers) != 2 {
			t.Errorf("Expected 2 FileManagers, got %d", len(env.FileManagers))
		}
		if env.FileManagers[0] != mockFileManager {
			t.Error("First FileManager not correct")
		}
		if env.FileManagers[1] != mockFileManager2 {
			t.Error("Second FileManager not correct")
		}
	})

	t.Run("should handle nil file manager", func(t *testing.T) {
		env := NewEnvironment(nil, nil)

		env.AddFileManager(nil)

		if len(env.FileManagers) != 1 {
			t.Errorf("Expected 1 FileManager, got %d", len(env.FileManagers))
		}
		if env.FileManagers[0] != nil {
			t.Error("Expected nil FileManager")
		}
	})
}

func TestClearFileManagers(t *testing.T) {
	t.Run("should clear empty file managers list", func(t *testing.T) {
		env := NewEnvironment(nil, nil)

		env.ClearFileManagers()

		if len(env.FileManagers) != 0 {
			t.Errorf("Expected empty FileManagers, got %d", len(env.FileManagers))
		}
	})

	t.Run("should clear file managers list with items", func(t *testing.T) {
		mockFileManager := &MockFileManager{}
		mockFileManager2 := &MockFileManager{}
		env := NewEnvironment(nil, []FileManager{mockFileManager, mockFileManager2})
		if len(env.FileManagers) != 2 {
			t.Error("Expected 2 FileManagers initially")
		}

		env.ClearFileManagers()

		if len(env.FileManagers) != 0 {
			t.Errorf("Expected empty FileManagers, got %d", len(env.FileManagers))
		}
	})

	t.Run("should allow adding file managers after clearing", func(t *testing.T) {
		mockFileManager := &MockFileManager{}
		env := NewEnvironment(nil, []FileManager{mockFileManager})

		env.ClearFileManagers()
		mockFileManager2 := &MockFileManager{}
		env.AddFileManager(mockFileManager2)

		if len(env.FileManagers) != 1 {
			t.Errorf("Expected 1 FileManager, got %d", len(env.FileManagers))
		}
		if env.FileManagers[0] != mockFileManager2 {
			t.Error("FileManager not correct after clearing and adding")
		}
	})
}

func TestIntegrationTests(t *testing.T) {
	t.Run("should work with complete workflow", func(t *testing.T) {
		externalEnv := map[string]any{
			"encodeBase64": func() string { return "encoded" },
			"mimeLookup":   func() string { return "text/css" },
		}

		mockFileManager := &MockFileManager{}
		env := NewEnvironment(externalEnv, []FileManager{mockFileManager})

		// Test external environment functions
		if env.EncodeBase64() != "encoded" {
			t.Error("EncodeBase64 not working correctly")
		}
		if env.MimeLookup() != "text/css" {
			t.Error("MimeLookup not working correctly")
		}

		// Test file manager operations
		mockFileManager2 := &MockFileManager{}
		env.AddFileManager(mockFileManager2)
		if len(env.FileManagers) != 2 {
			t.Error("Expected 2 FileManagers after adding")
		}

		// Test getFileManager
		mockFileManager2.SupportsFunc = func(string, string, map[string]any, map[string]any) bool { return true }
		result := env.GetFileManager("test.less", "/path", map[string]any{}, map[string]any{}, false)
		if result != mockFileManager2 {
			t.Error("GetFileManager should return mockFileManager2")
		}

		// Test clear
		env.ClearFileManagers()
		if len(env.FileManagers) != 0 {
			t.Error("Expected empty FileManagers after clearing")
		}
	})

	t.Run("should handle complex plugin manager scenario", func(t *testing.T) {
		pluginFileManager1 := &MockFileManager{
			SupportsFunc: func(string, string, map[string]any, map[string]any) bool { return false },
		}
		pluginFileManager2 := &MockFileManager{
			SupportsFunc: func(string, string, map[string]any, map[string]any) bool { return true },
		}
		complexPluginManager := &MockPluginManager{
			GetFileManagersFunc: func() []FileManager {
				return []FileManager{pluginFileManager1, pluginFileManager2}
			},
		}

		mockFileManager := &MockFileManager{
			SupportsFunc: func(string, string, map[string]any, map[string]any) bool { return true },
		}
		env := NewEnvironment(nil, []FileManager{mockFileManager})

		options := map[string]any{"pluginManager": complexPluginManager}
		result := env.GetFileManager("test.less", "/path", options, map[string]any{}, false)

		// Should return the plugin file manager that supports (checked first due to reverse order)
		if result != pluginFileManager2 {
			t.Error("Should return pluginFileManager2")
		}
	})
}